pkg/kubernetes/events/events.go

```diff
@@ -0,0 +1,145 @@
+package events
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/record"
+)
+
+// EventRecorder is a type used to record Kubernetes events
+type EventRecorder struct {
+	recorder record.EventRecorder
+	object   runtime.Object
+	watcher  watch.Interface
+}
+
+var (
+	once                 sync.Once
+	genericEventRecorder *EventRecorder
+)
+
+const (
+	// eventSource is the name of the event source which generates Kubernetes events
+	eventSource = "osm-controller"
+
+	// fatalEventPrefix is the prefix used with fatal event reasons which is used to identify Fatal events.
+	fatalEventPrefix = "Fatal"
+
+	// fatalEventWaitTimeout is the timeout period for waiting on a fatal event
+	fatalEventWaitTimeout = 10 * time.Second
+)
+
+// NewEventRecorder returns a new EventRecorder object
+func NewEventRecorder(object runtime.Object, kubeClient kubernetes.Interface, namespace string) *EventRecorder {
+	recorder, watcher := eventRecorder(kubeClient, namespace)
+	return &EventRecorder{
+		recorder: recorder,
+		watcher:  watcher,
+		object:   object,
+	}
+}
+
+// GenericEventRecorder is a singleton that returns a generic EventRecorder type.
+// The EventRecorder returned needs to be explicitly initialized by calling the 'Initialize' method on the object
+func GenericEventRecorder() *EventRecorder {
+	once.Do(func() {
+		genericEventRecorder = &EventRecorder{}
+	})
+
+	return genericEventRecorder
+}
+
+// eventRecorder returns an EventRecorder that can be used to post Kubernetes events, and its watcher
+func eventRecorder(kubeClient kubernetes.Interface, namespace string) (record.EventRecorder, watch.Interface) {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(log.Trace().Msgf)
+	eventBroadcaster.StartRecordingToSink(
+		&typedcorev1.EventSinkImpl{
+			Interface: kubeClient.CoreV1().Events(namespace)})
+	recorder := eventBroadcaster.NewRecorder(
+		scheme.Scheme,
+		corev1.EventSource{Component: eventSource})
+	watcher, err := kubeClient.CoreV1().Events(namespace).Watch(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		log.Error().Err(err).Msg("error getting watcher")
+	}
```

Comment:
What if we separated this into another function?
Is there a reason for the EventRecorder and watcher to be in the same fn?

Reply:
The `EventRecorder` object comprises the recorder as well as watcher, since `newEventRecorder` returns a new object, made sense to initialize the object in one function.